kernel/src/time/wait.rs

```diff
@@ -9,141 +9,195 @@ use crate::prelude::*;
-    /// this function will return `ETIME` error.
-    fn wait_until_or_timeout<F, R>(&self, cond: F, timeout: &Duration) -> Result<R>
-    where
-        F: FnMut() -> Option<R>,
-    {
-        let timer_builder = TimerBuilder::new(timeout);
-        self.wait_until_or_timer_timeout(cond, &timer_builder)
-    }
-
-    /// Similar to [`WaitTimeout::wait_until_or_timeout`].
+    /// Waits until some condition returns `Some(_)` or a given timeout is reached.
+    ///
+    /// # Errors
+    ///
+    /// If the condition does not become `Some(_)` before the timeout is reached, this function
+    /// will return an error with [`ETIME`].
     ///
-    /// The difference is that the timeout of this method is against the specified clock,
-    /// which is defined in `timer_builder`.
-    fn wait_until_or_timer_timeout<F, R>(&self, cond: F, timer_builder: &TimerBuilder) -> Result<R>
+    /// [`ETIME`]: crate::error::Errno::ETIME
+    fn wait_until_or_timeout<'a, F, T, R>(&self, mut cond: F, timeout: T) -> Result<R>
     where
         F: FnMut() -> Option<R>,
+        T: Into<TimeoutExt<'a>>,
     {
-        self.wait_until_or_timer_timeout_cancelled(cond, || Ok(()), timer_builder)
+        let timeout = timeout.into();
+        let timeout_inner = match timeout.check_expired() {
+            Ok(inner) => inner,
+            Err(err) => return cond().ok_or(err),
+        };
+
+        self.wait_until_or_timeout_cancelled(cond, || Ok(()), timeout_inner)
     }
 
-    /// Waits until some condition returns `Some(_)`, or be cancelled due to
-    /// reaching the timeout or the inputted cancel condition. If the condition
-    /// does not becomes `Some(_)` before the timeout is reached or `cancel_cond`
-    /// returns `Err`, this function will return corresponding `Err`.
+    /// Waits until some condition returns `Some(_)` or is cancelled by the timeout or cancel
+    /// condition.
+    ///
+    /// # Errors
+    ///
+    /// If the condition does not become `Some(_)` before the cancellation, this function
+    /// will return:
+    ///  - an error with [`ETIME`] if the timeout is reached;
+    ///  - the error returned by the cancel condition if the cancel condition returns `Err(_)`.
     #[doc(hidden)]
-    fn wait_until_or_timer_timeout_cancelled<F, R, FCancel>(
+    fn wait_until_or_timeout_cancelled<F, R, FCancel>(
         &self,
         cond: F,
         cancel_cond: FCancel,
-        timer_builder: &TimerBuilder,
+        timeout: Option<&ManagedTimeout>,
     ) -> Result<R>
     where
         F: FnMut() -> Option<R>,
         FCancel: Fn() -> Result<()>;
 }
 
-/// A helper structure to build timers from a specified timeout and timer manager.
-pub struct TimerBuilder<'a> {
-    timeout: Timeout,
-    timer_manager: &'a Arc<TimerManager>,
+/// A timeout with extended semantics.
+pub enum TimeoutExt<'a> {
+    /// The timeout will never fire.
+    Never,
+    /// The timeout is expired or will immediately fire.
+    Immediate,
```

Comment:
Introducing `Immediate` appears to be unnecessary and creates some inconsistencies. For instance, in `impl From<&Duration> for TimeoutExt<'_>`, the value is checked to see if it is zero. However, in `impl<'a> From<ManagedTimeout<'a>> for TimeoutExt<'a>`, no such check is performed. In fact, these checks could be deferred until `check_timeout` is called. Therefore, there seems to be no justification for introducing `Immediate` in this enum.